core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -362,12 +365,18 @@ class Partition(val topic: String,
    * 1. Partition ISR changed
    * 2. Any replica's LEO changed
    *
+   * We only increase HW if HW is smaller than LEO of all replicas whose lag <= replicaLagTimeMaxMs.
+   * This means that if a replica does not lag much behind leader and the replica's LEO is smaller than HW, HW will
+   * wait for this replica to catch up so that this replica can be added to ISR set.
```

Comment:
Tweak the comment a bit to the following. What do you think?

The HW is determined by the smallest log end offset among all replicas that are in sync or are considered caught-up. This way, if a replica is considered caught-up, but its log end offset is smaller than HW, we will wait for this replica to catch up to the HW before advancing the HW. This helps the situation when the ISR only includes the leader replica and a follower tries to catch up. If we don't wait for the follower when advancing the HW, the follower's log end offset may keep falling behind the HW (determined by the leader's log end offset) and therefore will never be added to ISR.